docs/blog/posts/pydantic.md

```diff
@@ -0,0 +1,303 @@
+
+## Introducing Tools Calling
+
+While tool calling was originally designed to make calls to external APIs using JSON schema, its real value lies in allowing us to specify the desired output. Fortunately, `Pydantic` provides utilities for generating a JSON schema and supports nested structures, which would be difficult to describe in plain text.
+
+In this example, instead of describing the desired output in plain text, we can simply provide the JSON schema for the `Packages` class, which includes a list of `Package` objects.
+
+As an exercise, try prompting the model to generate this prompt without using Pydantic!
+
+??? note "Example without Pydantic"
+
+    Heres the same example as below without using pydantic's schema generation
+
+    ```python
+    resp = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=[
+            {
+                "role": "user",
+                "content": "Pydantic and FastAPI?",
+            },
+        ],
+        tools=[
+            {
+                "type": "function",
+                "function": {
+                    "name": "Requirements",
+                    "description": "A list of packages and their first authors.",
+                    "parameters": {
+                        "$defs": {
+                            "Package": {
+                                "properties": {
+                                    "name": {"title": "Name", "type": "string"},
+                                    "author": {"title": "Author", "type": "string"},
+                                },
+                                "required": ["name", "author"],
+                                "title": "Package",
+                                "type": "object",
+                            }
+                        },
+                        "properties": {
+                            "packages": {
+                                "items": {"$ref": "#/$defs/Package"},
+                                "title": "Packages",
+                                "type": "array",
+                            }
+                        },
+                        "required": ["packages"],
+                        "title": "Packages",
+                        "type": "object",
+                    },
+                },
+            }
+        ],
+        tool_choice={
+            "type": "function",
+            "function": {"name": "Requirements"},
+        },
+    )
+
+    resp = json.loads(resp.choices[0].message.tool_calls[0].function.arguments)
+    ```
+
+Now, notice in this example that the prompts we use contain purely the data we want, where the tools and tools choice now capture the schemas we want to output. This speration of concerns makes it much easier organize the 'data' and the 'description' of the data that we want back out.
```

Comment:
Also I think this works better if you move this line below the code block — I look at the code and kind of see what's going on, and this then tells me what to think about what I just read.

The point is that I think, at least for me, it's easier for me to process the "commentary" (i.e., "[such and such] makes it much easier to [do something]") if I see the concrete thing first and can reflect on whether I agree with the commentary while reading it (note: I do. Lol).

I also think it could make sense to keep the first sentence above the code, and just move the second sentence below, since the first sentence draws my attention to something concrete (rather than editorializing, which I think in this case I'd rather consider after seeing the code).